### PR TITLE
feat: add externalIds support for visitor identification

### PR DESCRIPTION
**Pull request by ricardogarim**
## Proposed changes (including videos or screenshots)
This PR adds support for external identifiers (`externalIds`) on livechat visitors, enabling WhatsApp BSUID support, username display, and progressive visitor enrichment.

## Issue(s)
- [CORE-1818](https://rocketchat.atlassian.net/browse/CORE-1818) and [CORE-1835](https://rocketchat.atlassian.net/browse/CORE-1835)
- [ADR-0157](https://github.com/RocketChat/adrs/tree/main/adr/0157)

## Steps to test or reproduce
1. Apps-engine flow:
    - Use a WhatsApp app that calls `livechatCreator.resolveVisitor({ userId: '<BSUID>', username: '@johndoe' }, { phone: '<phone>' })`
    - Note: apps do NOT pass `source` - the bridge automatically injects the appId
    - Verify visitor is created/found with `externalIds` populated
2. Progressive enrichment:
    - Create a visitor via phone only (no `externalId`)
    - Call `resolveVisitor` with the same phone but new `externalId`
    - Verify the visitor record is enriched with the `externalId`
3. Email fallback:
    - Create a visitor via email only (no `externalId`)
    - Call `resolveVisitor` with `{ email: '<email>' }` as contactData
    - Verify the visitor record is enriched with the `externalId`
4. UI display:
    - Open a contact's info panel
    - Check the channels section displays username alongside phone
5. Index verification:
    - Run the explain query in the collapsible section below
    - Verify `winningPlan` uses `IXSCAN` with `indexName: 'externalIds.source_1_externalIds.userId_1'`

## Further comments
- **`externalIds` on Visitor, not Contact**: external identifiers are channel-specific (e.g., WhatsApp BSUID), so they belong on the visitor. Contacts reference visitors via `channels[].visitor.visitorId`, and the UI fetches visitor data to display usernames.
- **Progressive enrichment**: when a visitor exists by phone or email but not by `externalId`, we enrich the existing record rather than creating a duplicate. This handles the migration case where visitors were created before BSUID support.
- **`ResolveVisitorContactData` union type**: the `resolveVisitor()` API accepts `{ phone: string } | { email: string }` for flexible fallback, making it extensible for future contact data types without breaking changes.
- **Array structure with compound index**: `externalIds` is an array of `{ source, userId, username? }` objects using the [attribute pattern](https://www.mongodb.com/blog/post/building-with-patterns-the-attribute-pattern). A compound multikey index on `{ "externalIds.source": 1, "externalIds.userId": 1 }` enables efficient `$elemMatch` queries. This was chosen over a Record/wildcard index for ~1.5-2× smaller index size and tighter index scans.
- **App isolation**: apps do NOT pass `source` - the bridge automatically injects the appId. This ensures each app can only access its own external identifiers. The method signature uses `Omit<IVisitorExternalIdentifier, 'source'>` to enforce this at the type level.

<details>
  <summary>Index creation evidence</summary

```text
mongosh localhost:3001/meteor
meteor [direct: primary] meteor> db.rocketchat_livechat_visitor.find({
|     externalIds: {
|       $elemMatch: {
|         source: "a1b2c3d4-e5f6-4890-8bcd-ef1234567890",
|         userId: "bsuid-123"
|       }
|     }
|   }).explain("executionStats")
```

```javascript
{
  explainVersion: '1',
  queryPlanner: {
    namespace: 'meteor.rocketchat_livechat_visitor',
    indexFilterSet: false,
    parsedQuery: {
      externalIds: {
        '$elemMatch': {
          '$and': [
            {
              source: { '$eq': 'a1b2c3d4-e5f6-4890-8bcd-ef1234567890' }
            },
            { userId: { '$eq': 'bsuid-123' } }
          ]
        }
      }
    },
    queryHash: '5AE3527E',
    planCacheKey: '93651E13',
    maxIndexedOrSolutionsReached: false,
    maxIndexedAndSolutionsReached: false,
    maxScansToExplodeReached: false,
    winningPlan: {
      stage: 'FETCH',
      filter: {
        externalIds: {
          '$elemMatch': {
            '$and': [
              {
                source: { '$eq': 'a1b2c3d4-e5f6-4890-8bcd-ef1234567890' }
              },
              { userId: { '$eq': 'bsuid-123' } }
            ]
          }
        }
      },
      inputStage: {
        stage: 'IXSCAN',
        keyPattern: { 'externalIds.source': 1, 'externalIds.userId': 1 },
        indexName: 'externalIds.source_1_externalIds.userId_1',
        isMultiKey: true,
        multiKeyPaths: {
          'externalIds.source': [ 'externalIds' ],
          'externalIds.userId': [ 'externalIds' ]
        },
        isUnique: false,
        isSparse: false,
        isPartial: false,
        indexVersion: 2,
        direction: 'forward',
        indexBounds: {
          'externalIds.source': [
            '["a1b2c3d4-e5f6-4890-8bcd-ef1234567890", "a1b2c3d4-e5f6-4890-8bcd-ef1234567890"]'
          ],
          'externalIds.userId': [ '["bsuid-123", "bsuid-123"]' ]
        }
      }
    },
    rejectedPlans: [
      {
        stage: 'FETCH',
        filter: {
          externalIds: {
            '$elemMatch': {
              '$and': [
                { userId: { '$eq': 'bsuid-123' } },
                {
                  source: { '$eq': 'a1b2c3d4-e5f6-4890-8bcd-ef1234567890' }
                }
              ]
            }
          }
        },
        inputStage: {
          stage: 'IXSCAN',
          keyPattern: { '$_path': 1, 'externalIds.userId': 1 },
          indexName: 'externalIds.$**_1',
          isMultiKey: true,
          multiKeyPaths: { '$_path': [], 'externalIds.userId': [ 'externalIds' ] },
          isUnique: false,
          isSparse: false,
          isPartial: false,
          indexVersion: 2,
          direction: 'forward',
          indexBounds: {
            '$_path': [ '["externalIds.userId", "externalIds.userId"]' ],
            'externalIds.userId': [ '["bsuid-123", "bsuid-123"]' ]
          }
        }
      },
      {
        stage: 'FETCH',
        filter: {
          externalIds: {
            '$elemMatch': {
              '$and': [
                {
                  source: { '$eq': 'a1b2c3d4-e5f6-4890-8bcd-ef1234567890' }
                },
                { userId: { '$eq': 'bsuid-123' } }
              ]
            }
          }
        },
        inputStage: {
          stage: 'IXSCAN',
          keyPattern: { '$_path': 1, 'externalIds.source': 1 },
          indexName: 'externalIds.$**_1',
          isMultiKey: true,
          multiKeyPaths: { '$_path': [], 'externalIds.source': [ 'externalIds' ] },
          isUnique: false,
          isSparse: false,
          isPartial: false,
          indexVersion: 2,
          direction: 'forward',
          indexBounds: {
            '$_path': [ '["externalIds.source", "externalIds.source"]' ],
            'externalIds.source': [
              '["a1b2c3d4-e5f6-4890-8bcd-ef1234567890", "a1b2c3d4-e5f6-4890-8bcd-ef1234567890"]'
            ]
          }
        }
      }
    ]
  },
  executionStats: {
    executionSuccess: true,
    nReturned: 1,
    executionTimeMillis: 0,
    totalKeysExamined: 1,
    totalDocsExamined: 1,
    executionStages: {
      stage: 'FETCH',
      filter: {
        externalIds: {
          '$elemMatch': {
            '$and': [
              {
                source: { '$eq': 'a1b2c3d4-e5f6-4890-8bcd-ef1234567890' }
              },
              { userId: { '$eq': 'bsuid-123' } }
            ]
          }
        }
      },
      nReturned: 1,
      executionTimeMillisEstimate: 0,
      works: 3,
      advanced: 1,
      needTime: 0,
      needYield: 0,
      saveState: 0,
      restoreState: 0,
      isEOF: 1,
      docsExamined: 1,
      alreadyHasObj: 0,
      inputStage: {
        stage: 'IXSCAN',
        nReturned: 1,
        executionTimeMillisEstimate: 0,
        works: 2,
        advanced: 1,
        needTime: 0,
        needYield: 0,
        saveState: 0,
        restoreState: 0,
        isEOF: 1,
        keyPattern: { 'externalIds.source': 1, 'externalIds.userId': 1 },
        indexName: 'externalIds.source_1_externalIds.userId_1',
        isMultiKey: true,
        multiKeyPaths: {
          'externalIds.source': [ 'externalIds' ],
          'externalIds.userId': [ 'externalIds' ]
        },
        isUnique: false,
        isSparse: false,
        isPartial: false,
        indexVersion: 2,
        direction: 'forward',
        indexBounds: {
          'externalIds.source': [
            '["a1b2c3d4-e5f6-4890-8bcd-ef1234567890", "a1b2c3d4-e5f6-4890-8bcd-ef1234567890"]'
          ],
          'externalIds.userId': [ '["bsuid-123", "bsuid-123"]' ]
        },
        keysExamined: 1,
        seeks: 1,
        dupsTested: 1,
        dupsDropped: 0
      }
    }
  },
  command: {
    find: 'rocketchat_livechat_visitor',
    filter: {
      externalIds: {
        '$elemMatch': {
          source: 'a1b2c3d4-e5f6-4890-8bcd-ef1234567890',
          userId: 'bsuid-123'
        }
      }
    },
    '$db': 'meteor'
  },
  serverInfo: {
    host: 'Ricardos-MacBook-Pro.local',
    port: 3001,
    version: '7.0.16',
    gitVersion: '18b949444cfdaa88e30b0e10243bc18268251c1f'
  },
  serverParameters: {
    internalQueryFacetBufferSizeBytes: 104857600,
    internalQueryFacetMaxOutputDocSizeBytes: 104857600,
    internalLookupStageIntermediateDocumentMaxSizeBytes: 104857600,
    internalDocumentSourceGroupMaxMemoryBytes: 104857600,
    internalQueryMaxBlockingSortMemoryUsageBytes: 104857600,
    internalQueryProhibitBlockingMergeOnMongoS: 0,
    internalQueryMaxAddToSetBytes: 104857600,
    internalDocumentSourceSetWindowFieldsMaxMemoryBytes: 104857600,
    internalQueryFrameworkControl: 'trySbeRestricted'
  },
  ok: 1,
  '$clusterTime': {
    clusterTime: Timestamp({ t: 1773875387, i: 1 }),
    signature: {
      hash: Binary.createFromBase64('AAAAAAAAAAAAAAAAAAAAAAAAAAA=', 0),
      keyId: Long('0')
    }
  },
  operationTime: Timestamp({ t: 1773875387, i: 1 })
}
```
</details>

[CORE-1818]: https://rocketchat.atlassian.net/browse/CORE-1818?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CORE-1835]: https://rocketchat.atlassian.net/browse/CORE-1835?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **New Features**
  * Support external visitor identifiers across Livechat, including storing and propagating external IDs during visitor creation and lookup.
  * Resolve visitors by external identifier with optional phone/email fallback to match or enrich visitors.
  * Omnichannel UI surfaces show external identifier info alongside visitor details (improved display of username/phone).
* **Tests**
  * Added unit and end-to-end tests covering external ID resolution, phone/email fallback, enrichment, overwrite, and edge cases.

### CODE REVIEW

**Comment by ricardogarim** [comment]
@coderabbitai review

File: apps/meteor/client/views/omnichannel/directory/components/ContactField.tsx [hunk 1]
```diff
@@ -39,17 +38,19 @@ const ContactField = ({ contact, room }: ContactFieldProps) => {
 	}
 
 	const {
-		visitor: { username, name },
+		visitor: { username, name, phone },
 	} = data;
 
 	const displayName = name || username;
+	const phoneNumber = phone?.[0]?.phoneNumber;
+	const shortName = username && phoneNumber && username !== phoneNumber ? `${username} · ${phoneNumber}` : username || phoneNumber;
```
[the remark this answers is not in the document]
Comment (ricardogarim) [reply apps/meteor/client/views/omnichannel/directory/components/ContactField.tsx]: Intentional. `ContactField` shows the visitor's internal system identifier, while `ContactInfoChannelsItem` shows per-channel external identifiers. Different contexts, different purposes - no change needed.

File: packages/apps-engine/src/server/bridges/LivechatBridge.ts [hunk 2]
```diff
@@ -95,6 +95,12 @@ export abstract class LivechatBridge extends BaseBridge {
 		}
 	}
 
+	public async doResolveVisitor(externalId: IVisitorExternalIdentifier, phone: string | undefined, appId: string): Promise<IVisitor | undefined> {
+		if (this.hasReadPermission(appId, 'livechat-visitor')) {
+			return this.resolveVisitor(externalId, phone, appId);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ricardogarim) [reply packages/apps-engine/src/server/bridges/LivechatBridge.ts]: Fixed on [ed3f7fa](https://github.com/RocketChat/Rocket.Chat/pull/39535/commits/ed3f7fae197a2cc27d7fde3b8ac776256eb7a628). Changed `doResolveVisitor` to require `livechat-visitor.write` instead of read, since the method can enrich the visitor via `addExternalId`. However, I'm not 100% certain this is the ideal approach - an alternative would be to split into a pure lookup (read) and a separate enrichment method (write). Open to feedback on this design decision.

File: packages/models/src/models/LivechatVisitors.ts [hunk 3]
```diff
@@ -49,6 +50,25 @@ export class LivechatVisitorsRaw extends BaseRaw<ILivechatVisitor> implements IL
...
+	}
+
+	addExternalId(_id: string, externalId: IVisitorExternalIdentifier): Promise<UpdateResult> {
+		return this.updateOne({ _id }, [
+			{
+				$set: {
+					externalIds: { $setUnion: [{ $ifNull: ['$externalIds', []] }, [externalId]] },
+				},
+			},
+		]);
```
[the remark this answers is not in the document]
Comment (ricardogarim) [reply packages/models/src/models/LivechatVisitors.ts]: Fixed on [34d5267](https://github.com/RocketChat/Rocket.Chat/pull/39535/commits/34d526719e7560a16632011a89a4c0eb451c3405). We can have just one item per source inside `externalIds`.

File: apps/meteor/app/apps/server/bridges/livechat.ts [hunk 4]
```diff
@@ -235,6 +236,7 @@ export class AppLivechatBridge extends LivechatBridge {
 			id: visitor.id,
 			...(visitor.phone?.length && { phone: { number: visitor.phone[0].phoneNumber } }),
 			...(visitor.visitorEmails?.length && { email: visitor.visitorEmails[0].address }),
+			...(visitor.externalIds?.length && { externalIds: visitor.externalIds }),
```
[the remark this answers is not in the document]
Comment (ricardogarim) [reply apps/meteor/app/apps/server/bridges/livechat.ts]: Intentional. Added a deprecation note on [42d1cd3](https://github.com/RocketChat/Rocket.Chat/pull/39535/commits/42d1cd3c42589dc401ae16f8f2eaf979686f87d0). The idea is to make users to call the new `createAndReturnVisitor()` method.

File: apps/meteor/app/livechat/server/lib/resolveVisitor.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+export async function resolveVisitor({ externalId, phone }: ResolveVisitorParams): Promise<ILivechatVisitor | null> {
+	const visitorByExternalId = await LivechatVisitors.findOneByExternalId(externalId.source, externalId.userId);
+	if (visitorByExternalId) {
+		return visitorByExternalId;
+	}
+
+	if (phone) {
+		const visitorByPhone = await LivechatVisitors.findOneVisitorByPhone(phone);
```
Comment (d-gubert) [inline apps/meteor/app/livechat/server/lib/resolveVisitor.ts]: If you can replace this with a `findOneAndUpdate` call it would be cool, since you save a db roundtrip. Not sure if you can optionally `$addToSet` though.

Comment (d-gubert) [reply apps/meteor/app/livechat/server/lib/resolveVisitor.ts]: Checked the update query and it seems it already guards against potential duplication, so maybe the `findOneAndUpdate` is feasible

File: packages/core-typings/src/ILivechatVisitor.ts [hunk 6]
```diff
@@ -26,6 +32,7 @@ export interface ILivechatVisitor extends IRocketChatRecord {
 	ip?: string;
 	host?: string;
 	visitorEmails?: IVisitorEmail[];
+	externalIds?: IVisitorExternalIdentifier[];
```
Comment (d-gubert) [inline packages/core-typings/src/ILivechatVisitor.ts:35]: Have you considered making this an embedded doc instead of an array? For instance

```
externalIds?: Record<IVisitorExternalIdentifier['source'], IVisitorExternalIdentifier>;
```

Sounds like it would be easier to manage checking source existence and updating the record

Comment (ricardogarim) [reply packages/core-typings/src/ILivechatVisitor.ts:35]: [a8375ed](https://github.com/RocketChat/Rocket.Chat/pull/39535/commits/a8375ed7a239de9c94d80fc53c711041041d4b95) :D as we discussed on RC channel.

**Comment by ricardogarim** [comment]
@coderabbitai review

File: apps/meteor/app/livechat/server/lib/resolveVisitor.ts (hunk 5 above)
Comment (ricardogarim) [reply apps/meteor/app/livechat/server/lib/resolveVisitor.ts]: Good catch. [52cd3a9](https://github.com/RocketChat/Rocket.Chat/pull/39535/commits/52cd3a9650120e7e669c8d9b28f844d19eb4e484).

**Review by cardoso — DISMISSED** [review]

**Review by tassoevan — DISMISSED** [review]

**Review by d-gubert — CHANGES_REQUESTED** [review]
Please address the bots' comments on the change set.

Additionally, please add tests for the new Apps-Engine API - you can follow the pattern seen in this PR https://github.com/RocketChat/Rocket.Chat/pull/38374/changes#diff-3edecfa2dd0bd2aec99f5a218abb5d57fcd0c85d434fd7a87136e9f1e3461232

**Comment by ricardogarim** [comment]
> Additionally, please add tests for the new Apps-Engine API - you can follow the pattern seen in this PR https://github.com/RocketChat/Rocket.Chat/pull/38374/changes#diff-3edecfa2dd0bd2aec99f5a218abb5d57fcd0c85d434fd7a87136e9f1e3461232

Added as a separate PR: https://github.com/RocketChat/Rocket.Chat/pull/39657.

**Review by KevLehman — COMMENTED** [review]
Curious: what happens to visitors that are already registered without an external ID when a new message comes in?

File: packages/models/src/models/LivechatVisitors.ts [hunk 7]
```diff
@@ -49,6 +50,24 @@ export class LivechatVisitorsRaw extends BaseRaw<ILivechatVisitor> implements IL
 		return this.findOne(query);
 	}
 
+	findOneVisitorByPhoneAndAddExternalId(
```
Comment (KevLehman) [inline packages/models/src/models/LivechatVisitors.ts]: On this naming: i think the most important part of the query is an update, so the method should be update too.

File: packages/omni-core/src/visitor/create.ts [hunk 8]
```diff
@@ -24,13 +25,21 @@ export const registerGuest = makeFunction(
 
 		logger.debug({ msg: 'New incoming conversation', id, token });
 
-		const visitorDataToUpdate: Partial<ILivechatVisitor> & { userAgent?: string; ip?: string; host?: string } = {
+		const visitorDataToUpdate: Partial<ILivechatVisitor> & { userAgent?: string; ip?: string; host?: string } & Record<string, unknown> = {
```
Comment (KevLehman) [inline packages/omni-core/src/visitor/create.ts]: This typing seems too broad, could be something like:
```
Record<`externalIds.${string}`, string>
```
Cause we know externalId would be an string (unless we change to be an array-like)

File: apps/meteor/app/livechat/server/lib/resolveVisitor.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+	phone?: string;
+};
+
+export async function resolveVisitor({ source, externalId, phone }: ResolveVisitorParams): Promise<ILivechatVisitor | null> {
+	const visitorByExternalId = await LivechatVisitors.findOneByExternalId(source, externalId.userId);
+	if (visitorByExternalId) {
+		return visitorByExternalId;
+	}
+
+	if (phone) {
```
Comment (KevLehman) [inline apps/meteor/app/livechat/server/lib/resolveVisitor.ts]: if `externalId` is null/undef it would fail the check above, reaching here, and then it would add `{ appId: undefined }` to the visitor object.

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: packages/omni-core/src/visitor/create.ts (hunk 8 above)
Comment (ricardogarim) [reply packages/omni-core/src/visitor/create.ts]: [f586b4d](https://github.com/RocketChat/Rocket.Chat/pull/39535/commits/f586b4d6ddb680035fa0584f588cfd9cb90c6c8b)

File: .changeset/forty-dolphins-check.md [hunk 10]
```diff
@@ -0,0 +1,10 @@
+---
+'@rocket.chat/model-typings': patch
+'@rocket.chat/core-typings': minor
+'@rocket.chat/apps-engine': minor
+'@rocket.chat/omni-core': patch
+'@rocket.chat/models': patch
+'@rocket.chat/meteor': patch
```
Comment (d-gubert) [inline .changeset/forty-dolphins-check.md]:
```suggestion
'@rocket.chat/model-typings': minor
'@rocket.chat/core-typings': minor
'@rocket.chat/apps-engine': minor
'@rocket.chat/omni-core': minor
'@rocket.chat/models': minor
'@rocket.chat/meteor': minor
```

Since this is a "feat", everybody gets a minor

File: packages/omni-core/src/visitor/create.ts [hunk 11]
```diff
@@ -24,13 +25,24 @@ export const registerGuest = makeFunction(
...
+				IVisitorExternalIdentifier
+			> = {
 			token,
 			status,
 			...(phone?.number && { phone: [{ phoneNumber: phone.number }] }),
 			...(name && { name }),
 		};
 
+		// Use dot notation for `externalIds` to merge with existing entries
+		// instead of overwriting.
```
Comment (d-gubert) [inline packages/omni-core/src/visitor/create.ts]:
```suggestion
		// We "flatten" the `externalIds` from the parameters using dot notation so Mongo doesn't overwrite the whole property when updating the record
```

Making it clear in the comments that we're targeting Mongo's behavior. I got confused when I first saw this and didn't realize the "overwriting" we want to prevent was in the database context, not the javascript one 😬

File: apps/meteor/app/livechat/server/lib/resolveVisitor.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+	contactData?: ResolveVisitorContactData;
+};
+
+export async function resolveVisitor({ source, externalId, contactData }: ResolveVisitorParams): Promise<ILivechatVisitor | null> {
+	const visitorByExternalId = await LivechatVisitors.findOneByExternalId(source, externalId.userId);
+	if (visitorByExternalId) {
+		return visitorByExternalId;
+	}
+
+	if (contactData && (('phone' in contactData && contactData.phone) || ('email' in contactData && contactData.email))) {
```
Comment (d-gubert) [inline apps/meteor/app/livechat/server/lib/resolveVisitor.ts:18]: Won't block the merge over this, I think it's a nitpick - but I'll share anyway 😂 

```suggestion
	if (contactData?.phone || contactData?.email) {
```

Personally looks easier to understand and a more succint check

**Review by d-gubert — DISMISSED** [review]

**Review by d-gubert — DISMISSED** [review]

File: apps/meteor/app/livechat/server/lib/resolveVisitor.ts (hunk 12 above)
Comment (d-gubert) [reply apps/meteor/app/livechat/server/lib/resolveVisitor.ts:18]: Turns out the discriminated union burst my bubble :frowning_face:

File: packages/apps-engine/src/definition/livechat/IVisitor.ts [hunk 13]
```diff
@@ -14,4 +22,5 @@ export interface IVisitor {
 	activity?: string[];
 	customFields?: { [key: string]: any };
 	livechatData?: { [key: string]: any };
+	externalIds?: IVisitorExternalIdentifier[];
```
[the remark this answers is not in the document]
Comment (ricardogarim) [reply packages/apps-engine/src/definition/livechat/IVisitor.ts:25]: You're looking to the diff from last commit only. This is a new field being introduced, not a change to an existing API - and also there are no integrations using it yet.

**Comment by ricardogarim** [comment]
After discussing with Diego and exploring MongoDB index behavior, we switched `externalIds` back to an array structure instead of records.

Arrays with compound indexes only index the fields we query and allow filtering on multiple fields at once, while wildcard indexes on records index everything and can only filter one field per query.

Details in [ADR #157](https://adr.rocket.chat/0157).

**Review by d-gubert — DISMISSED** [review]

**Review by KevLehman — DISMISSED** [review]

File: apps/meteor/client/views/omnichannel/contactInfo/tabs/ContactInfoChannels/ContactInfoChannelsItem.tsx [hunk 14]
```diff
@@ -29,6 +30,19 @@ const ContactInfoChannelsItem = ({
...
+	const channelLabel = useMemo(() => {
+		const phone = getSourceLabel(details);
+		const externalId = details?.id ? visitorData?.externalIds?.find((e) => e.source === details.id) : undefined;
+		const username = externalId?.username;
+
+		if (username && phone) {
+			return `${username} - ${phone}`;
+		}
+		return username || phone;
+	}, [visitorData?.externalIds, details, getSourceLabel]);
```
Comment (gabriellsh) [inline apps/meteor/client/views/omnichannel/contactInfo/tabs/ContactInfoChannels/ContactInfoChannelsItem.tsx:44]: From what I read on the ADR, shouldn't Email be a fallback too?

Comment (ricardogarim) [reply apps/meteor/client/views/omnichannel/contactInfo/tabs/ContactInfoChannels/ContactInfoChannelsItem.tsx:44]: From what we discussed, email fallback is currently intended only for backend purposes - mainly to allow the new `resolveVisitor` to work in a more generic way without introducing breaking changes to the interface in the future.

On the frontend side, for now, I think we’re comfortable not relying on it. wdyt?

Comment (gabriellsh) [reply apps/meteor/client/views/omnichannel/contactInfo/tabs/ContactInfoChannels/ContactInfoChannelsItem.tsx:44]: Yep. I thought it would be something like this, I just wanted to question to make sure we weren't missing something. Thanks for the explanation!

**Review by gabriellsh — DISMISSED** [review]

File: apps/meteor/client/views/omnichannel/directory/components/ContactField.tsx (hunk 1 above)
Comment (ggazzo) [inline apps/meteor/client/views/omnichannel/directory/components/ContactField.tsx]: is this a  actually a short name :) ?  it seems to me that is more like a compound name, also It doesn't seem related to the scope of the pr.


also I would do a bit different ,Perhaps it's just a different way of thinking but her emy suggestion:  	` const shortName = [...new Set([username, phoneNumber ])].filter(Boolean).join(' ')	`

**Review by ggazzo — CHANGES_REQUESTED** [review]

File: packages/core-typings/src/ILivechatVisitor.ts [hunk 15]
```diff
@@ -14,6 +14,12 @@ export interface IVisitorEmail {
 	address: string;
 }
 
+export interface IVisitorExternalIdentifier {
+	source: string;
```
Comment (sampaiodiego) [inline packages/core-typings/src/ILivechatVisitor.ts]: `source` is optional on the "apps-engine" type at https://github.com/RocketChat/Rocket.Chat/pull/39535/changes#diff-682f105a369ae6b44fadc871598bde8b29fa3b1d952320383e3112b590f42987R5

not sure which one is correct

File: packages/apps-engine/src/definition/livechat/IVisitor.ts [hunk 16]
```diff
@@ -1,6 +1,14 @@
 import type { IVisitorEmail } from './IVisitorEmail';
 import type { IVisitorPhone } from './IVisitorPhone';
 
+export interface IVisitorExternalIdentifier {
+	source?: string;
```
Comment (sampaiodiego) [inline packages/apps-engine/src/definition/livechat/IVisitor.ts]: I guess this isn't optional
```suggestion
	source: string;
```

File: apps/meteor/tests/unit/app/livechat/server/lib/resolveVisitor.spec.ts
Comment (sampaiodiego) [inline apps/meteor/tests/unit/app/livechat/server/lib/resolveVisitor.spec.ts:1]: not sure about the real value of theses tests, since the `resolveVisitor` function is just calling the two models methods that are being mocked

**Review by sampaiodiego — CHANGES_REQUESTED** [review]

File: ee/packages/pdf-worker/src/strategies/ChatTranscript.ts [hunk 17]
```diff
@@ -28,6 +28,7 @@ export class ChatTranscript implements IStrategy {
 						visitor: requestData.visitor
 							? {
 									...requestData.visitor,
+									externalIds: undefined,
```
Comment (sampaiodiego) [inline ee/packages/pdf-worker/src/strategies/ChatTranscript.ts]: what is the purpose of this change? I might be overlooking but I couldn't figure it out why it is needed

File: apps/meteor/client/views/omnichannel/directory/components/ContactField.tsx (hunk 1 above)
Comment (ricardogarim) [reply apps/meteor/client/views/omnichannel/directory/components/ContactField.tsx]: fair. changes added on https://github.com/RocketChat/Rocket.Chat/pull/39535/commits/3b82c913f0565be4deabdc3fb723fed192fd10cd.

**Review by sampaiodiego — APPROVED** [review]

**Review by ggazzo — APPROVED** [review]
